sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSet.scala

```diff
@@ -74,49 +83,37 @@ class ExpressionSet protected(
 
   protected def remove(e: Expression): Unit = {
     if (e.deterministic) {
-      baseSet --= baseSet.filter(_ == e.canonicalized)
-      originals --= originals.filter(_.canonicalized == e.canonicalized)
+      baseSet.retain(_ != e.canonicalized)
```

Comment:
`baseSet.retain(_ != e.canonicalized)` only needs to traverse the set once, much better than the previous implementation.